### PR TITLE
ADP-91-User-Object-Creation

### PR DESCRIPTION
**Pull request by nick-mcguffin**
---
name: Pull request
about: Request to merge code
title: 'ADP-91-User-Object-Creation'
labels: ''
assignees: @nick-mcguffin

---

**PR Type** *(Select all that apply)*
  - [x] Feature 
  - [ ] Bug fix
  - [x] Improvement 
  - [ ] Other

**Feature Description**

New users can create a new account - Educator, Partner or Student - and it will be persisted to the database successfully.

**Submission Checklist** *(do not remove)*

Use the following as a guide to ensure your PR is complete and acceptable:

- [x] You are NOT attempting to merge into the `master` branch
- [x] You are requesting to merge your branch into the `development` branch or one of its sub-branches
- [x] Your code compiles and runs without errors (on as many operating systems as practicable
- [x] Code has sufficient Javadoc comments (giving examples where needed)
- [x] Code is clear and readable, being sure to use meaningful names and effective whitespace
- [x] Appropriate [naming convention](https://en.wikipedia.org/wiki/Naming_convention_(programming)) is used according to the language used
- [x] Code duplication is minimized through use of polymorphism and inheritance
- [x] Dependencies have been checked for vulnerabilities
- [x] There is sufficient unit tests for the code and all tests pass
- [x] Unit tests utilize mock data/tools and don't write data to any databases
- [x] There is no logic in any views, use service/controller methods as helpers 
- [x] Use spaces between method parameters `[1, 2, 3]` not `[1,2,3]` and around operators `a + b` not `a+b`
- [x] All work has been thoroughly checked for grammatical and spelling errors
- [x] Documentation is to be clear, concise, and written in English
- [x] Changes are documented in a `README.md` if applicable

### CODE REVIEW

**Comment by nick-mcguffin** [comment]
@nathansnow1981 Not sure whether you want to make any changes to the entity constructors. I have created new constructors to persist the new user accounts to the database so you can see my changes before we figure out what to do with user profiles? 

Also something to think about - we need to compare a username entered in a registration form against our database of usernames to prevent new users from attempting to submit a username that is already used as WILMA will throw an exception and be sad. I have added a comment regarding this in [ADP-92 Validate User Input](https://appdevaces.atlassian.net/browse/ADP-92".

File: service/src/main/java/com/wilma/service/UserRegistrationService.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
+package com.wilma.service;
+
```
Comment (nathansnow1981) [inline service/src/main/java/com/wilma/service/UserRegistrationService.java:2]:
```suggestion
import lombok.extern.slf4j.Slf4j;
```

File: service/src/main/java/com/wilma/service/UserRegistrationService.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+import org.springframework.stereotype.Service;
+
+import com.wilma.entity.dto.USER_TYPE;
+import com.wilma.entity.dto.UserDTO;
+import com.wilma.entity.users.Educator;
+import com.wilma.entity.users.Partner;
+import com.wilma.entity.users.Student;
+import com.wilma.entity.users.UserAccount;
+import com.wilma.repository.UserAccountRepository;
+
```
Comment (nathansnow1981) [inline service/src/main/java/com/wilma/service/UserRegistrationService.java:14]:
```suggestion

@Slf4j
```

File: service/src/main/java/com/wilma/service/UserRegistrationService.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        else if(userDTO.getUserType().equalsIgnoreCase(USER_TYPE.STUDENT.name())) return createStudentFromUserDTO(userDTO);
+        else throw new UnsupportedOperationException("User type '" + userDTO.getUserType() + "' could not be identified.\nSupported types are 'educator', 'partner', and 'student'");
+    }
+
+    private Educator createEducatorFromUserDTO(UserDTO userDTO) {
+        Educator educator = new Educator(userDTO.getUserId(),
+        userDTO.getUsername(), passwordEncoder.encode(userDTO.getPassword()), userDTO.getEmail(),
+        null, false, false,
+        false, false, null, null, userDTO.getDiscipline(), userDTO.getStaffId(), userDTO.getFirstName(), userDTO.getLastName(), userDTO.getContactNumber()); 
+        System.err.print("A new " + userDTO.getUserType() + " has been added to the database.");
```
Comment (nathansnow1981) [inline service/src/main/java/com/wilma/service/UserRegistrationService.java:35]:
```suggestion
log.info("A new {} has been added to the database", userDTO.getUserType());
```

File: service/src/main/java/com/wilma/service/UserRegistrationService.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        return userRepository.save(educator);
+    }
+
+    public Partner createPartnerFromUserDTO(UserDTO userDTO){
+        Partner partner = new Partner (userDTO.getUserId(),
+                        userDTO.getUsername(), passwordEncoder.encode(userDTO.getPassword()), userDTO.getEmail(),
+                        null, false, false,
+                        false, false, null,  userDTO.getBusinessName(),
+                        userDTO.getFirstName(), userDTO.getLastName(), userDTO.getContactNumber(), userDTO.getAbn());
+            System.err.print("A new " + userDTO.getUserType() + " has been added to the database.");
```
Comment (nathansnow1981) [inline service/src/main/java/com/wilma/service/UserRegistrationService.java:45]:
```suggestion
log.info("A new {} has been added to the database", userDTO.getUserType());
```

File: service/src/main/java/com/wilma/service/UserRegistrationService.java [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+            return userRepository.save(partner);
+        }
+
+    private Student createStudentFromUserDTO(UserDTO userDTO) {
+        Student student = new Student(userDTO.getUserId(),
+        userDTO.getUsername(), passwordEncoder.encode(userDTO.getPassword()), userDTO.getEmail(),
+        null, false, false,
+        false, false, null, null, userDTO.getDiscipline(), userDTO.getStudentId(), userDTO.getFirstName(), userDTO.getLastName(), userDTO.getContactNumber());
+        System.err.print("A new " + userDTO.getUserType() + " has been added to the database.");
```
Comment (nathansnow1981) [inline service/src/main/java/com/wilma/service/UserRegistrationService.java:54]:
```suggestion
log.info("A new {} has been added to the database", userDTO.getUserType());
```

**Comment by nathansnow1981** [comment]
Yes that's very true re: `usernames`.....
Hopefully this is something that can be implemented with simple validation, but if not we'll have to sort something else out.
I'll keep an eye out for potential solutions.... Nice catch 👌

Also, these entities look great, and the new constructor was definitely the least invasive solution

**Review by nathansnow1981 — APPROVED** [review]
Nicely done 👍

**Pull request merged**